Implement gui

david-ribeiro-sumup opened the pull request:
Implemented a simple GUI to demo the terminal API. Refactored code to be usable both as a script and on the UI. Updated README.md with instructions.

david-ribeiro-sumup committed:
Started implementation of GUI for terminal API.

Using fresh as a fullstack framework.
Created very basic UI to start implementing its functionality.
Refactored the scripts to run transactions into their own directory.
Created deno tasks to run transaction scripts.
Extracted ts types into its own directory.

david-ribeiro-sumup committed:
Refactored transactions code to a class.

Transactions code now works on UI and CLI.
Refactored transaction code to a TransactionRunner class which also includes a ping and cancel methods.
Refactored logging so we can later log to the UI.

david-ribeiro-sumup committed:
Implement logger based on react state.

Created a logger for the UI that populates a state array with all log entries.
Disabling buttons and inputs when a transaction is currently active.
Created a simple UI to render the logs.

david-ribeiro-sumup committed:
Improving css.

david-ribeiro-sumup committed:
Improved css. Implemented config validation.

david-ribeiro-sumup committed:
Fixed bug in logger. Implemented delay between requests.

chad-lung-fivestars commented:
I have some feedback for this. We can review that today.

chad-lung-fivestars commented on terminal-api/utils.ts:
```diff
@@ -0,0 +1,130 @@
+  timeout = 120000
+): Promise<Response> {
+  let response: Response;
+
```
The "delay" should be put in here and removed from the other spots. We want the delay to slow all calls using httpRequest()

david-ribeiro-sumup committed:
Implemented request and response logging.

Refactored logging to include a more complex data structure that supports requests and responses.
Found a JSON inspector component that works with fresh and preact.
Using it to display complex JSON objects.
Added coloring to responses and flow outcome.

david-ribeiro-sumup committed:
Moved delay code to httpRequest.

david-ribeiro-sumup committed:
Created one page for config and another for running flows.

UI to run flows is shown after successfully testing a connection to the terminal.

david-ribeiro-sumup committed:
Modal to confirm discount and payment flows.

After connection is established with the terminal the customer service is polled in background.
Added a button to clear logs at any time.
A modal is shown when the customer selects a discount to confirm it.
Reimplemented functions to manage transactions.
Created a transaction status monitor that continuously checks the transaction state.

david-ribeiro-sumup committed:
Terminating monitor on completed transaction.

An extra request was being made caused by a re-render.

david-ribeiro-sumup committed:
Small fixes and improvements to the UI.

Fixed rendering of 0s caused conditions resulting in non boolean values.
Added a status area with the current state of the terminal API and possible actions.
Added a dedicated area to show the latest customer endpoint information.
Added another well with the current transaction information.

david-ribeiro-sumup committed:
Clearing transaction state upon completion to solve bug.

david-ribeiro-sumup committed:
Added delay between requests to customer endpoint monitor. Made requests cancellable from outside.

david-ribeiro-sumup committed:
Implemented debounce for delay configuration input.

chad-lung-fivestars commented:
The latest update looks really good. The one thing I noticed is that you can't cancel. You should be able to cancel anytime from the customer checking in up until the payment is started.

chad-lung-fivestars commented:
FYI - I was only able to apply a reward once. After that I couldn't select cash/credit/other in my next transaction.

david-ribeiro-sumup committed:
Fixed bug preventing discounts from being applied in different transactions.

david-ribeiro-sumup commented:
> The latest update looks really good. The one thing I noticed is that you can't cancel. You should be able to cancel anytime from the customer checking in up until the payment is started.

The API to cancel needs the pos-checkout-id, so I guess I can't cancel until I have one. Am I missing something?

david-ribeiro-sumup commented:
> FYI - I was only able to apply a reward once. After that I couldn't select cash/credit/other in my next transaction.

I think I fixed it in the commit I just made

david-ribeiro-sumup committed:
Solved duplicated modal bug.

Cleaning approved discount when device becomes idle instead of finishing transaction.

david-ribeiro-sumup committed:
Allowing cancel at any time.

chad-lung-fivestars reviewed (approved):
👍  This looks great now and cancel is working as expected. Please merge when you get a chance.
